### ISSUE TITLE
mergeData and returnSamples fail when sample_id is not globally unique

### ISSUE THREAD

**Issue by lwaldron** [waldronlab/curatedMetagenomicData#244]
**Describe the bug**

sample_id is only guaranteed to be unique within a `study_name`, but it can be repeated across different studies. mergeData and returnSamples fail when the user tries to merge datasets containing some of the same sample_id values. It should be possible to merge these studies, for example by pre-pending study_name to the sample_id to create globally unique identifiers.

**To Reproduce**

``` r
suppressPackageStartupMessages(library(curatedMetagenomicData))
df <- sampleMetadata[sampleMetadata$study_name %in% c("LeChatelierE_2013", "NielsenHB_2014"),]
suppressMessages(returnSamples(df, dataType = "rel"))
#> Error: colnames/sample_id values are not unique
```

<sup>Created on 2021-06-08 by the [reprex package](https://reprex.tidyverse.org) (v2.0.0)</sup>

<details style="margin-bottom:10px;">
<summary>
Session info
</summary>

``` r
sessioninfo::session_info()
#> ─ Session info ───────────────────────────────────────────────────────────────
#>  setting  value                       
#>  version  R version 4.1.0 (2021-05-18)
#>  os       Ubuntu 18.04.5 LTS          
#>  system   x86_64, linux-gnu           
#>  ui       X11                         
#>  language (EN)                        
#>  collate  en_US.UTF-8                 
#>  ctype    en_US.UTF-8                 
#>  tz       America/New_York            
#>  date     2021-06-08                  
#> 
#> ─ Packages ───────────────────────────────────────────────────────────────────
#>  package                  * version  date       lib
#>  AnnotationDbi              1.54.0   2021-05-19 [1]
#>  AnnotationHub              3.0.0    2021-05-19 [1]
#>  ape                        5.5      2021-04-25 [1]
#>  assertthat                 0.2.1    2019-03-21 [1]
#>  backports                  1.2.1    2020-12-09 [1]
#>  Biobase                  * 2.52.0   2021-05-19 [1]
#>  BiocFileCache              2.0.0    2021-05-19 [1]
#>  BiocGenerics             * 0.38.0   2021-05-19 [1]
#>  BiocManager                1.30.15  2021-05-11 [1]
#>  BiocParallel               1.26.0   2021-05-19 [1]
#>  BiocVersion                3.13.1   2021-03-19 [1]
#>  Biostrings               * 2.60.0   2021-05-19 [1]
#>  bit                        4.0.4    2020-08-04 [1]
#>  bit64                      4.0.5    2020-08-30 [1]
#>  bitops                     1.0-7    2021-04-24 [1]
#>  blob                       1.2.1    2020-01-20 [1]
#>  cachem                     1.0.5    2021-05-15 [1]
#>  cli                        2.5.0    2021-04-26 [1]
#>  crayon                     1.4.1    2021-02-08 [1]
#>  curatedMetagenomicData   * 3.1.2    2021-06-07 [1]
#>  curl                       4.3.1    2021-04-30 [1]
#>  DBI                        1.1.1    2021-01-15 [1]
#>  dbplyr                     2.1.1    2021-04-06 [1]
#>  DelayedArray               0.18.0   2021-05-19 [1]
#>  digest                     0.6.27   2020-10-24 [1]
#>  dplyr                      1.0.6    2021-05-05 [1]
#>  ellipsis                   0.3.2    2021-04-29 [1]
#>  evaluate                   0.14     2019-05-28 [1]
#>  ExperimentHub              2.0.0    2021-05-19 [1]
#>  fansi                      0.5.0    2021-05-25 [1]
#>  fastmap                    1.1.0    2021-01-25 [1]
#>  filelock                   1.0.2    2018-10-05 [1]
#>  fs                         1.5.0    2020-07-31 [1]
#>  generics                   0.1.0    2020-10-31 [1]
#>  GenomeInfoDb             * 1.28.0   2021-05-19 [1]
#>  GenomeInfoDbData           1.2.6    2021-05-25 [1]
#>  GenomicRanges            * 1.44.0   2021-05-19 [1]
#>  glue                       1.4.2    2020-08-27 [1]
#>  highr                      0.9      2021-04-16 [1]
#>  htmltools                  0.5.1.1  2021-01-22 [1]
#>  httpuv                     1.6.1    2021-05-07 [1]
#>  httr                       1.4.2    2020-07-20 [1]
#>  interactiveDisplayBase     1.30.0   2021-05-19 [1]
#>  IRanges                  * 2.26.0   2021-05-19 [1]
#>  KEGGREST                   1.32.0   2021-05-19 [1]
#>  knitr                      1.33     2021-04-24 [1]
#>  later                      1.2.0    2021-04-23 [1]
#>  lattice                    0.20-44  2021-05-02 [2]
#>  lifecycle                  1.0.0    2021-02-15 [1]
#>  magrittr                   2.0.1    2020-11-17 [1]
#>  Matrix                     1.3-3    2021-05-04 [2]
#>  MatrixGenerics           * 1.4.0    2021-05-19 [1]
#>  matrixStats              * 0.59.0   2021-06-01 [1]
#>  memoise                    2.0.0    2021-01-26 [1]
#>  mime                       0.10     2021-02-13 [1]
#>  nlme                       3.1-152  2021-02-04 [2]
#>  pillar                     1.6.1    2021-05-16 [1]
#>  pkgconfig                  2.0.3    2019-09-22 [1]
#>  png                        0.1-7    2013-12-03 [1]
#>  promises                   1.2.0.1  2021-02-11 [1]
#>  purrr                      0.3.4    2020-04-17 [1]
#>  R6                         2.5.0    2020-10-28 [1]
#>  rappdirs                   0.3.3    2021-01-31 [1]
#>  Rcpp                       1.0.6    2021-01-15 [1]
#>  RCurl                      1.98-1.3 2021-03-16 [1]
#>  reprex                     2.0.0    2021-04-02 [1]
#>  rlang                      0.4.11   2021-04-30 [1]
#>  rmarkdown                  2.8      2021-05-07 [1]
#>  RSQLite                    2.2.7    2021-04-22 [1]
#>  S4Vectors                * 0.30.0   2021-05-19 [1]
#>  sessioninfo                1.1.1    2018-11-05 [1]
#>  shiny                      1.6.0    2021-01-25 [1]
#>  SingleCellExperiment     * 1.14.1   2021-05-21 [1]
#>  stringi                    1.6.2    2021-05-17 [1]
#>  stringr                    1.4.0    2019-02-10 [1]
#>  styler                     1.4.1    2021-03-30 [1]
#>  SummarizedExperiment     * 1.22.0   2021-05-19 [1]
#>  tibble                     3.1.2    2021-05-16 [1]
#>  tidyr                      1.1.3    2021-03-03 [1]
#>  tidyselect                 1.1.1    2021-04-30 [1]
#>  TreeSummarizedExperiment * 2.0.0    2021-05-19 [1]
#>  utf8                       1.2.1    2021-03-12 [1]
#>  vctrs                      0.3.8    2021-04-29 [1]
#>  withr                      2.4.2    2021-04-18 [1]
#>  xfun                       0.23     2021-05-15 [1]
#>  xtable                     1.8-4    2019-04-21 [1]
#>  XVector                  * 0.32.0   2021-05-19 [1]
#>  yaml                       2.2.1    2020-02-01 [1]
#>  zlibbioc                   1.38.0   2021-05-19 [1]
#>  source                                            
#>  Bioconductor                                      
#>  Bioconductor                                      
#>  CRAN (R 4.1.0)                                    
#>  CRAN (R 4.1.0)                                    
#>  CRAN (R 4.1.0)                                    
#>  Bioconductor                                      
#>  Bioconductor                                      
#>  Bioconductor                                      
#>  CRAN (R 4.1.0)                                    
#>  Bioconductor                                      
#>  Bioconductor                                      
#>  Bioconductor                                      
#>  CRAN (R 4.1.0)                                    
#>  CRAN (R 4.1.0)                                    
#>  CRAN (R 4.1.0)                                    
#>  CRAN (R 4.1.0)                                    
#>  CRAN (R 4.1.0)                                    
#>  CRAN (R 4.1.0)                                    
#>  CRAN (R 4.1.0)                                    
#>  Github (waldronlab/curatedMetagenomicData@c702763)
#>  CRAN (R 4.1.0)                                    
#>  CRAN (R 4.1.0)                                    
#>  CRAN (R 4.1.0)                                    
#>  Bioconductor                                      
#>  CRAN (R 4.1.0)                                    
#>  CRAN (R 4.1.0)                                    
#>  CRAN (R 4.1.0)                                    
#>  CRAN (R 4.1.0)                                    
#>  Bioconductor                                      
#>  CRAN (R 4.1.0)                                    
#>  CRAN (R 4.1.0)                                    
#>  CRAN (R 4.1.0)                                    
#>  CRAN (R 4.1.0)                                    
#>  CRAN (R 4.1.0)                                    
#>  Bioconductor                                      
#>  Bioconductor                                      
#>  Bioconductor                                      
#>  CRAN (R 4.1.0)                                    
#>  CRAN (R 4.1.0)                                    
#>  CRAN (R 4.1.0)                                    
#>  CRAN (R 4.1.0)                                    
#>  CRAN (R 4.1.0)                                    
#>  Bioconductor                                      
#>  Bioconductor                                      
#>  Bioconductor                                      
#>  CRAN (R 4.1.0)                                    
#>  CRAN (R 4.1.0)                                    
#>  CRAN (R 4.1.0)                                    
#>  CRAN (R 4.1.0)                                    
#>  CRAN (R 4.1.0)                                    
#>  CRAN (R 4.0.5)                                    
#>  Bioconductor                                      
#>  CRAN (R 4.1.0)                                    
#>  CRAN (R 4.1.0)                                    
#>  CRAN (R 4.1.0)                                    
#>  CRAN (R 4.0.3)                                    
#>  CRAN (R 4.1.0)                                    
#>  CRAN (R 4.1.0)                                    
#>  CRAN (R 4.1.0)                                    
#>  CRAN (R 4.1.0)                                    
#>  CRAN (R 4.1.0)                                    
#>  CRAN (R 4.1.0)                                    
#>  CRAN (R 4.1.0)                                    
#>  CRAN (R 4.1.0)                                    
#>  CRAN (R 4.1.0)                                    
#>  CRAN (R 4.1.0)                                    
#>  CRAN (R 4.1.0)                                    
#>  CRAN (R 4.1.0)                                    
#>  CRAN (R 4.1.0)                                    
#>  Bioconductor                                      
#>  CRAN (R 4.1.0)                                    
#>  CRAN (R 4.1.0)                                    
#>  Bioconductor                                      
#>  CRAN (R 4.1.0)                                    
#>  CRAN (R 4.1.0)                                    
#>  CRAN (R 4.1.0)                                    
#>  Bioconductor                                      
#>  CRAN (R 4.1.0)                                    
#>  CRAN (R 4.1.0)                                    
#>  CRAN (R 4.1.0)                                    
#>  Bioconductor                                      
#>  CRAN (R 4.1.0)                                    
#>  CRAN (R 4.1.0)                                    
#>  CRAN (R 4.1.0)                                    
#>  CRAN (R 4.1.0)                                    
#>  CRAN (R 4.1.0)                                    
#>  Bioconductor                                      
#>  CRAN (R 4.1.0)                                    
#>  Bioconductor                                      
#> 
#> [1] /usr/local/lib/R/site-library
#> [2] /usr/lib/R/library
```

</details>

**Expected behavior**

The previous cMD merge functions dealt with this by using `paste(study_name, sample_id, sep=":")` as sample names in the merged dataset. Another, less intrusive approach could be to pre-pend study name only to the non-unique sample_id values, which currently would only affect samples from two studies (but may affect more in the future).

**Additional context**

Curators use sample_id as found in the original publication / SRA submission, rather than changing e.g. to enforce globally-unique sample_id, to enable referring back to the original publication or SRA submission for more information about the sample. Duplicated sample_id may occur by chance, or may indicate actual re-use of specimens across different studies. These are currently the only two studies with duplicated sample_id.

**Comment by schifferl** [comment]
I think to solve the problem for the long-term the `study_name:sample_id` strategy is pretty reasonable. This would be:
```
LeChatelierE_2013:Sample_A
NielsenHB_2014:Sample_A
...
```
It might take me a few weeks to get to this, but I will and I'll go with the stated strategy.

**Comment by lwaldron** [comment]
Thanks @schifferl. By the way, I noticed that this error occurs also if the offending duplicates are removed from the `returnSamples` input:

``` r
suppressPackageStartupMessages(library(curatedMetagenomicData))
df <- sampleMetadata[sampleMetadata$study_name %in% c("LeChatelierE_2013", "NielsenHB_2014"),]
dups <-
  df$study_name == "NielsenHB_2014" &
  grepl(pattern = "^MH", x = df$sample_id)
df <- df[!dups, ]
summary(isUnique(df$sample_id))
#>    Mode    TRUE 
#> logical     511
suppressMessages(returnSamples(df, dataType = "rel"))
#> Error: colnames/sample_id values are not unique
```

<sup>Created on 2021-06-09 by the [reprex package](https://reprex.tidyverse.org) (v2.0.0)</sup>

<details style="margin-bottom:10px;">
<summary>
Session info
</summary>

``` r
sessioninfo::session_info()
#> ─ Session info ───────────────────────────────────────────────────────────────
#>  setting  value                       
#>  version  R version 4.1.0 (2021-05-18)
#>  os       Ubuntu 18.04.5 LTS          
#>  system   x86_64, linux-gnu           
#>  ui       X11                         
#>  language (EN)                        
#>  collate  en_US.UTF-8                 
#>  ctype    en_US.UTF-8                 
#>  tz       America/New_York            
#>  date     2021-06-09                  
#> 
#> ─ Packages ───────────────────────────────────────────────────────────────────
#>  package                  * version  date       lib
#>  AnnotationDbi              1.54.0   2021-05-19 [1]
#>  AnnotationHub              3.0.0    2021-05-19 [1]
#>  ape                        5.5      2021-04-25 [1]
#>  assertthat                 0.2.1    2019-03-21 [1]
#>  backports                  1.2.1    2020-12-09 [1]
#>  Biobase                  * 2.52.0   2021-05-19 [1]
#>  BiocFileCache              2.0.0    2021-05-19 [1]
#>  BiocGenerics             * 0.38.0   2021-05-19 [1]
#>  BiocManager                1.30.15  2021-05-11 [1]
#>  BiocParallel               1.26.0   2021-05-19 [1]
#>  BiocVersion                3.13.1   2021-03-19 [1]
#>  Biostrings               * 2.60.0   2021-05-19 [1]
#>  bit                        4.0.4    2020-08-04 [1]
#>  bit64                      4.0.5    2020-08-30 [1]
#>  bitops                     1.0-7    2021-04-24 [1]
#>  blob                       1.2.1    2020-01-20 [1]
#>  cachem                     1.0.5    2021-05-15 [1]
#>  cli                        2.5.0    2021-04-26 [1]
#>  crayon                     1.4.1    2021-02-08 [1]
#>  curatedMetagenomicData   * 3.1.2    2021-06-07 [1]
#>  curl                       4.3.1    2021-04-30 [1]
#>  DBI                        1.1.1    2021-01-15 [1]
#>  dbplyr                     2.1.1    2021-04-06 [1]
#>  DelayedArray               0.18.0   2021-05-19 [1]
#>  digest                     0.6.27   2020-10-24 [1]
#>  dplyr                      1.0.6    2021-05-05 [1]
#>  ellipsis                   0.3.2    2021-04-29 [1]
#>  evaluate                   0.14     2019-05-28 [1]
#>  ExperimentHub              2.0.0    2021-05-19 [1]
#>  fansi                      0.5.0    2021-05-25 [1]
#>  fastmap                    1.1.0    2021-01-25 [1]
#>  filelock                   1.0.2    2018-10-05 [1]
#>  fs                         1.5.0    2020-07-31 [1]
#>  generics                   0.1.0    2020-10-31 [1]
#>  GenomeInfoDb             * 1.28.0   2021-05-19 [1]
#>  GenomeInfoDbData           1.2.6    2021-05-25 [1]
#>  GenomicRanges            * 1.44.0   2021-05-19 [1]
#>  glue                       1.4.2    2020-08-27 [1]
#>  highr                      0.9      2021-04-16 [1]
#>  htmltools                  0.5.1.1  2021-01-22 [1]
#>  httpuv                     1.6.1    2021-05-07 [1]
#>  httr                       1.4.2    2020-07-20 [1]
#>  interactiveDisplayBase     1.30.0   2021-05-19 [1]
#>  IRanges                  * 2.26.0   2021-05-19 [1]
#>  KEGGREST                   1.32.0   2021-05-19 [1]
#>  knitr                      1.33     2021-04-24 [1]
#>  later                      1.2.0    2021-04-23 [1]
#>  lattice                    0.20-44  2021-05-02 [2]
#>  lifecycle                  1.0.0    2021-02-15 [1]
#>  magrittr                   2.0.1    2020-11-17 [1]
#>  Matrix                     1.3-4    2021-06-01 [2]
#>  MatrixGenerics           * 1.4.0    2021-05-19 [1]
#>  matrixStats              * 0.59.0   2021-06-01 [1]
#>  memoise                    2.0.0    2021-01-26 [1]
#>  mime                       0.10     2021-02-13 [1]
#>  nlme                       3.1-152  2021-02-04 [2]
#>  pillar                     1.6.1    2021-05-16 [1]
#>  pkgconfig                  2.0.3    2019-09-22 [1]
#>  png                        0.1-7    2013-12-03 [1]
#>  promises                   1.2.0.1  2021-02-11 [1]
#>  purrr                      0.3.4    2020-04-17 [1]
#>  R6                         2.5.0    2020-10-28 [1]
#>  rappdirs                   0.3.3    2021-01-31 [1]
#>  Rcpp                       1.0.6    2021-01-15 [1]
#>  RCurl                      1.98-1.3 2021-03-16 [1]
#>  reprex                     2.0.0    2021-04-02 [1]
#>  rlang                      0.4.11   2021-04-30 [1]
#>  rmarkdown                  2.8      2021-05-07 [1]
#>  RSQLite                    2.2.7    2021-04-22 [1]
#>  S4Vectors                * 0.30.0   2021-05-19 [1]
#>  sessioninfo                1.1.1    2018-11-05 [1]
#>  shiny                      1.6.0    2021-01-25 [1]
#>  SingleCellExperiment     * 1.14.1   2021-05-21 [1]
#>  stringi                    1.6.2    2021-05-17 [1]
#>  stringr                    1.4.0    2019-02-10 [1]
#>  styler                     1.4.1    2021-03-30 [1]
#>  SummarizedExperiment     * 1.22.0   2021-05-19 [1]
#>  tibble                     3.1.2    2021-05-16 [1]
#>  tidyr                      1.1.3    2021-03-03 [1]
#>  tidyselect                 1.1.1    2021-04-30 [1]
#>  TreeSummarizedExperiment * 2.0.0    2021-05-19 [1]
#>  utf8                       1.2.1    2021-03-12 [1]
#>  vctrs                      0.3.8    2021-04-29 [1]
#>  withr                      2.4.2    2021-04-18 [1]
#>  xfun                       0.23     2021-05-15 [1]
#>  xtable                     1.8-4    2019-04-21 [1]
#>  XVector                  * 0.32.0   2021-05-19 [1]
#>  yaml                       2.2.1    2020-02-01 [1]
#>  zlibbioc                   1.38.0   2021-05-19 [1]
#>  source                                            
#>  Bioconductor                                      
#>  Bioconductor                                      
#>  CRAN (R 4.1.0)                                    
#>  CRAN (R 4.1.0)                                    
#>  CRAN (R 4.1.0)                                    
#>  Bioconductor                                      
#>  Bioconductor                                      
#>  Bioconductor                                      
#>  CRAN (R 4.1.0)                                    
#>  Bioconductor                                      
#>  Bioconductor                                      
#>  Bioconductor                                      
#>  CRAN (R 4.1.0)                                    
#>  CRAN (R 4.1.0)                                    
#>  CRAN (R 4.1.0)                                    
#>  CRAN (R 4.1.0)                                    
#>  CRAN (R 4.1.0)                                    
#>  CRAN (R 4.1.0)                                    
#>  CRAN (R 4.1.0)                                    
#>  Github (waldronlab/curatedMetagenomicData@c702763)
#>  CRAN (R 4.1.0)                                    
#>  CRAN (R 4.1.0)                                    
#>  CRAN (R 4.1.0)                                    
#>  Bioconductor                                      
#>  CRAN (R 4.1.0)                                    
#>  CRAN (R 4.1.0)                                    
#>  CRAN (R 4.1.0)                                    
#>  CRAN (R 4.1.0)                                    
#>  Bioconductor                                      
#>  CRAN (R 4.1.0)                                    
#>  CRAN (R 4.1.0)                                    
#>  CRAN (R 4.1.0)                                    
#>  CRAN (R 4.1.0)                                    
#>  CRAN (R 4.1.0)                                    
#>  Bioconductor                                      
#>  Bioconductor                                      
#>  Bioconductor                                      
#>  CRAN (R 4.1.0)                                    
#>  CRAN (R 4.1.0)                                    
#>  CRAN (R 4.1.0)                                    
#>  CRAN (R 4.1.0)                                    
#>  CRAN (R 4.1.0)                                    
#>  Bioconductor                                      
#>  Bioconductor                                      
#>  Bioconductor                                      
#>  CRAN (R 4.1.0)                                    
#>  CRAN (R 4.1.0)                                    
#>  CRAN (R 4.1.0)                                    
#>  CRAN (R 4.1.0)                                    
#>  CRAN (R 4.1.0)                                    
#>  CRAN (R 4.1.0)                                    
#>  Bioconductor                                      
#>  CRAN (R 4.1.0)                                    
#>  CRAN (R 4.1.0)                                    
#>  CRAN (R 4.1.0)                                    
#>  CRAN (R 4.0.3)                                    
#>  CRAN (R 4.1.0)                                    
#>  CRAN (R 4.1.0)                                    
#>  CRAN (R 4.1.0)                                    
#>  CRAN (R 4.1.0)                                    
#>  CRAN (R 4.1.0)                                    
#>  CRAN (R 4.1.0)                                    
#>  CRAN (R 4.1.0)                                    
#>  CRAN (R 4.1.0)                                    
#>  CRAN (R 4.1.0)                                    
#>  CRAN (R 4.1.0)                                    
#>  CRAN (R 4.1.0)                                    
#>  CRAN (R 4.1.0)                                    
#>  CRAN (R 4.1.0)                                    
#>  Bioconductor                                      
#>  CRAN (R 4.1.0)                                    
#>  CRAN (R 4.1.0)                                    
#>  Bioconductor                                      
#>  CRAN (R 4.1.0)                                    
#>  CRAN (R 4.1.0)                                    
#>  CRAN (R 4.1.0)                                    
#>  Bioconductor                                      
#>  CRAN (R 4.1.0)                                    
#>  CRAN (R 4.1.0)                                    
#>  CRAN (R 4.1.0)                                    
#>  Bioconductor                                      
#>  CRAN (R 4.1.0)                                    
#>  CRAN (R 4.1.0)                                    
#>  CRAN (R 4.1.0)                                    
#>  CRAN (R 4.1.0)                                    
#>  CRAN (R 4.1.0)                                    
#>  Bioconductor                                      
#>  CRAN (R 4.1.0)                                    
#>  Bioconductor                                      
#> 
#> [1] /usr/local/lib/R/site-library
#> [2] /usr/lib/R/library
```

</details>

**Comment by schifferl** [comment]
Just an update @lwaldron, I found this was not entirely resolved by 8bcf3c0. I added another patch in e261e1f to fully resolve the issue.
